File: src/events/component_interaction/name_confirm.rs
use poise::serenity_prelude as serenity;

use serenity::*;

use crate::prelude::*;

pub async fn yes_clicked(
    context: &Context,
    component_interaction: &MessageComponentInteraction,
    bot_data: &BotData,
    name: String,
) -> BotResult<()> {
    // Try to find a matching user
    let verified_user = sqlx::query!(
        "select * from users
        where name = ? and registered_user_id is null
        order by kind
        limit 1",
        name,
    )
    .try_map(|row| VerifiedUser::from_columns(row.name, row.kind, row.registered_user_id))
    .fetch_optional(&bot_data.database_pool)
    .await?;

    match verified_user {
        // No matching name was found
        None => {
            // Update the user's registration state to failed and ask them to seek assistance
            reset_status(
                context,
                component_interaction,
                bot_data,
                RegistrationStatus::Failed,
                component_interaction.user.id,
                embeds::registration::name_error(bot_data.channels.support_channel_id),
            )
            .await?;

            // Alert mentors about the name error
            bot_data
                .channels
                .log_channel_id
                .send_message(&context.http, |message| {
                    message
                        .content(format!(
                            "<@&{}> <@&{}>",
                            bot_data.roles.hierarchy.mentor_role_id,
                            bot_data.roles.hierarchy.admin_role_id,
                        ))
                        .embed(embeds::logs::name_error(
                            component_interaction.user.id,
                            &name,
                            bot_data.channels.support_channel_id,
                        ))
                })
                .await
        },
        // A matching name was found and ask the user to confirm their kind
        Some(verified_user) => {
            // Confirm the user's kind
            request_kind_confirm(context, component_interaction, bot_data, verified_user).await?;

            // Log the name confirmation
            bot_data
                .channels
                .log_channel_id
                .send_message(&context.http, |message| {
                    message.embed(embeds::logs::name_confirmed(
                        component_interaction.user.id,
                        &name,
                    ))
                })
                .await
        },
    }?;

    Ok(())
}

pub async fn no_clicked(
    context: &Context,
    component_interaction: &MessageComponentInteraction,
    bot_data: &BotData,
) -> BotResult<()> {
    // Update the user's registration state to started and ask them to enter their name again
    reset_status(
        context,
        component_interaction,
        bot_data,
        RegistrationStatus::Started,
        component_interaction.user.id,
        embeds::registration::name_request(),
    )
    .await
}

async fn reset_status(
    context: &Context,
    component_interaction: &MessageComponentInteraction,
    bot_data: &BotData,
    status: RegistrationStatus,
    user_id: UserId,
    embed: impl FnOnce(&mut CreateEmbed) -> &mut CreateEmbed,
) -> BotResult<()> {
    // Update the user's registration state
    sqlx::query!(
        "update registrations
        set status = ?, name = null
        where user_id = ?",
        status.to_string(),
        user_id.0,
    )
    .execute(&bot_data.database_pool)
    .await?;

    // Display an embed
    component_interaction
        .create_followup_message(&context.http, |message| message.embed(embed))
        .await?;

    Ok(())
}

async fn request_kind_confirm(
    context: &Context,
    component_interaction: &MessageComponentInteraction,
    bot_data: &BotData,
    verified_user: VerifiedUser,
) -> BotResult<()> {
    // Update the user's registration state to name confirmed
    sqlx::query!(
        "update registrations
        set status = ?, kind = ?
        where user_id = ?", // NOTE: Name should have been set when name entered
        RegistrationStatus::KindFound(verified_user.name, verified_user.kind).to_string(),
        verified_user.kind.to_string(),
        component_interaction.user.id.0,
    )
    .execute(&bot_data.database_pool)
    .await?;

    // Ask the user to confirm their kind
    component_interaction
        .create_followup_message(&context.http, |message| {
            message
                .embed(embeds::registration::kind_confirm(verified_user.kind))
                .components(components::confirm_kind_buttons())
        })
        .await?;

    Ok(())
}

File: norris/events/component_interaction/name_confirm.py
from discord import Interaction
from sqlalchemy import select, update

from ...bot import Norris
from ...model import NameEntered, Registration, RegistrationStatus, VerifiedUser
from . import verify_registration_status


async def yes_clicked(interaction: Interaction, norris: Norris) -> None:
    # Defer response to give time for database queries
    await interaction.response.defer()

    # Check that the user has the correct state to click the button
    if not await verify_registration_status(interaction.user.id,
                                            RegistrationStatus.NAME_ENTERED,
                                            norris):
        return

    async with norris.database_engine.begin() as connection:
        # Retrieve the user's name
        result = await connection.execute(
            select(NameEntered)
            .where(Registration.user_id == interaction.user.id)
            .limit(1),
        )

        user_name = result.one().name  # NOTE: this should be a NameEntered

        # Try to find a matching user who is not registered
        result = await connection.execute(
            select(VerifiedUser)
            .where(VerifiedUser.name == user_name
                   and VerifiedUser.registered_user_id is None)
            .order_by(VerifiedUser.kind)
            .limit(1),
        )
        verified_user = result.one_or_none()

        from ...responses.components import KindConfirmView
        from ...responses import embeds

        if verified_user is None:
            # Update the user's registration state to failed
            await connection.execute(
                update(Registration)
                .values(status=RegistrationStatus.FAILED, name=None)
                .where(Registration.user_id == interaction.user.id),
            )

            # Ask the user to seek assistance
            await interaction.followup.send(
                embed=embeds.registration.name_error(
                    norris.channels.support_channel_id),
            )

            # Alert the mentors about no name being found
            await norris.get_channel(norris.channels.log_channel_id).send(
                embed=embeds.logs.name_error(
                    interaction.user.id,
                    norris.roles.hierarchy.mentor_role_id,
                    norris.channels.support_channel_id,
                    user_name
                ),
            )
        else:
            # Update the user's registration state to kind found
            await connection.execute(
                update(Registration)
                .values(status=RegistrationStatus.KIND_FOUND, kind=verified_user.kind)
                .where(Registration.user_id == interaction.user.id),
            )

            # Ask the user to confirm their kind
            await interaction.followup.send(
                embed=embeds.registration.kind_confirm(verified_user.kind),
                view=KindConfirmView(norris),
            )

            # Log the name confirmation
            await norris.get_channel(norris.channels.log_channel_id).send(
                embed=embeds.logs.name_confirmed(interaction.user.id,
                                                 user_name),
            )


async def no_clicked(interaction: Interaction, norris: Norris) -> None:
    # Defer response to give time for database queries
    await interaction.response.defer()

    # Check that the user has the correct state to click the button
    if not await verify_registration_status(interaction.user.id,
                                            RegistrationStatus.NAME_ENTERED,
                                            norris):
        return

    async with norris.database_engine.begin() as connection:
        # Update the user's registration state to started
        await connection.execute(
            update(Registration)
            .where(Registration.user_id == interaction.user.id)
            .values(status=RegistrationStatus.STARTED, name=None),
        )

    # NOTE: circular modules again
    from ...responses import embeds

    # Ask the user to enter their name
    await interaction.followup.send(embed=embeds.registration.name_request())
